### ISSUE TITLE
What happens if the server dies after db commit but before saving event to kafka?

### ISSUE THREAD

**Issue by thekalinga** [kbastani/event-sourcing-microservices-example#28]
Currently, friend-service tries to do an atomic operation across two different data sources. In reality, it is not atomic as the server can die after committing to db but before the message is persisted in kafka

Either we have to use 2-phase commit to guarantee consistency which is a performance killer (or) use CDC (or) add events too to the database in a single transaction to guarantee atomicity (or) use SAGA (or) use process manager

The current implementation leaves the whole system in inconsistent state

**Comment by kbastani** [comment]
We can indeed try this. Is this something you have already experimented with? If so, can you provide the logs and make a pull request that shows evidence that this does not work?

The tests themselves actually do test for this scenario. 

Here you see the error being thrown in the callback. This lambda function is equivalent of an application failure during an inflight transaction. 

The transaction itself is only finalized after the database operation is successful (no error in the callback) and the message is successfully sent to Kafka. Asynchronously, when the message is sent to Kafka, the receipt or acknowledgment is received back from Kafka (while still inside the transaction scope of the database), causing an atomic commit of the data and the Kafka message being received. This provides a 100% guarantee that an application-level dual-write cannot result in data loss, since an application failure after a database operation will cause a `ROLLBACK` to occur and a Kafka message not to be sent.

Here is the error callback defined as a filed in a test class of the `friend-service`.

https://github.com/kbastani/event-sourcing-microservices-example/blob/e67ffb9602424b95f32afabd44e176f5a905f39f/friend-service/src/test/java/io/example/domain/FriendServiceTests.java#L33

Here is the test execution that simulates the failure that you've described without sending a message to Kafka or committing the database transaction.

https://github.com/kbastani/event-sourcing-microservices-example/blob/e67ffb9602424b95f32afabd44e176f5a905f39f/friend-service/src/test/java/io/example/domain/FriendServiceTests.java#L111

I will leave this issue open for your response. If you are satisfied by the result, then please close this issue yourself.

**Comment by thekalinga** [comment]
@kbastani 

Put a `System.exit(0)` after these three lines & you will see that kafka has a message that database does not have

https://github.com/kbastani/event-sourcing-microservices-example/blob/e67ffb9602424b95f32afabd44e176f5a905f39f/friend-service/src/main/java/io/example/domain/friend/FriendController.java#L93-L95

**Comment by kbastani** [comment]
@thekalinga That does look like a problem, I must admit. Although, I believe that modifying the runtime code by adding a `System.exit(0)` is not a valid test for inflight transactions. I will take a look at this further and consult with @cer and @smaldini. Thanks for bringing this to my attention.

**Comment by kbastani** [comment]
@thekalinga I think I've found a suitable solution. I will use the `KafkaTemplate.inTransaction(callback)`:  https://docs.spring.io/spring-kafka/reference/html/_reference.html#chained-transaction-manager. This will cause both the writes to the database and to Kafka to be rolled back if there is an application crash during an in-flight transaction.

```java
boolean result = kafkaTemplate.executeInTransaction(t -> {
    t.sendDefault("foo", "bar");
    return true;
});
```

**Comment by thabach** [comment]
Hey @kbastani I believe you still suffer from the same inconsistency potential, rooted in crashing right after a successful deposit of the event in Kafka. AFAIR chaining the two separate transactions can not prevent from this.

**Comment by cer** [comment]
Yes. What will happen if the kafka transaction doesn't commit?

Also, you are coupling the service to the DB AND the message broker. Both must be up for a transaction to commit, which reduces availability.

**Comment by kbastani** [comment]
@cer If the Kafka transaction doesn't commit, then the parent database transaction will also fail. The key here is to understand the reactive workflow. These two systems are not a part of a serial transaction (2pc). 

Both the Kafka and database transactions finalize only when they are returned back and received by the end-user who has initiated the API request (since this is an end-to-end reactive example). When the user receives the final response in the stream, then the two transactions are committed. With reactor, the Mono<Friend> result represents the context of the database transaction and the Kafka transaction. When the transactions are both ready to be finalized, they are joined together as a part of the reactive flow. The transactional contexts are published back to the consumer as an event, which will then be committed `OnComplete`.

If the application crashes before the end-user received a response, then both Kafka and the database will be rolled back and the user will receive an error.

**Comment by kbastani** [comment]
@thabach Yes, it would seem so at first. However, both R2DBC and KafkaTemplate have an `executeInTransaction(...)` function that become a part of the reactive flow. If the server crashes before the response is returned back to the end-user client, then both the database and Kafka will be rolled back.

**Comment by kbastani** [comment]
Welcome to the new world of reactive transactions Chris. 😊

**Comment by kbastani** [comment]
@cer @thekalinga After looking into this, the current code works for any condition that would cause an error to occur during the dual-write, as pointed out by @thekalinga. To make this operation completely fault tolerant—in the case of an application fault during an in-flight transaction—there will need to be a status field on any entity participating in a dual-write.

For this case, the database transaction will first save the record to a versioned table with a key reference to any existing record in the live table. The versioned entities will describe the change in state that will be captured in the event. Since the data record will be committed to the database first, it will still not be operational until a message is sent to Kafka.

Since this is an eventually consistent architecture, the Kafka message must be sent once and only once. This table will contain a collection of versioned entities that represent the audit log and the change in state overtime. _Last write wins_ will be the isolation level. Which means, that any row in the versioned table that was not comitted, will become archived or deleted. 

Both the Kafka cluster and the Postgres cluster should be highly available and replicated to multiple AZs. The only risk of a inconsistency due to a fault is if the application handling the in-flight transaction suddenly crashes. If either the database or Kafka goes down, the transaction for both will fail _without question_. Resulting in an error to the front-end. In the rare case that the application crashes before or after a Kafka event is sent out, there will be no commit applied to the event, which will cause Kafka to rollback the change.

Since this system is eventually consistent. It is completely fine to reprocess the event from the Postgres database using another node. A worker thread will monitor the versioned table of the entity and will then retry dispatching the event to Kafka. Ideally, Kafka will not contain the state of the user object, but instead provide meta data and a callback to the system of record's API, with a version number that will be used when populating any third party system. 

The only way to ensure that there is 100% fault tolerance is to have a single durable external system have a record of the database write, but that data should not be operational until Kafka receives the event. Should a version number not exist in the system of record, the Kafka event will be thrown out and ignored by any third-party system looking to replay the event stream.

**Comment by thekalinga** [comment]
@kbastani 

> Welcome to the new world of reactive transactions Chris

Reactive is not a silver bullet against cross datasource transactions. It can fail exactly the same way both in imperative & declarative models

Also, TBH, the solution you proposed is quite error prone & not easy to get everything right

As far as I know, there are few off the shelf solutions based on db log streaming

1. Use CDC to tail database log using a framework like debezium (or)
2. Kafka connector that also can stream db log from postgres -> kafka

**Comment by adamflynn** [comment]
Does the operation really need to be atomic? Can you not write to Kafka first and then have a consumer of that event perform the database write?

This [article](https://www.confluent.io/blog/using-logs-to-build-a-solid-data-infrastructure-or-why-dual-writes-are-a-bad-idea/) has helped shape my thinking

**Comment by nurkiewicz** [comment]
I don't fully grasp the last suggestion by @kbastani, but indeed in principle, two transactions on separate systems (irrespective to their nature, guarantees and and availability) can't be atomic. You either need a third-party (like XA transaction manager) or _chain_ transaction. As long as you need to send to commit messages over network to independent systems, one of them can get lost. Any solution at the API level (reactive, callback-base, etc.) can overcome the issues with underlying network protocol.

The most straightforward solution is to either use Kafka as a source of truth and propagate it to secondary _views_ like databases or stream databases changes to Kafka. In both cases at-least-once semantics are needed.

As a matter of fact this is what relational database is doing all the time: storing data in transaction log (append-only data structure) and make sure all "views" (tables) reflect the change.

**Comment by trainings** [comment]
I totally agree with @nurkiewicz 
The solution is to use only one system/service as a Source of Truth.
And it can be different options:
- Kafka and implement smth like CQRS for filling DB
- PostgreSQL and logical replication functionality
- MsSQL and CDC functionality
etc.

**Comment by GuyPardon** [comment]
Coming back to the original question, and the remark: "2-phase commit to guarantee consistency which is a performance killer". That is not necessarily true: since you are already accessing 2 systems (both DBMS and Kafka), much of the overhead is already included in your case. 

We (atomikos) have been able to make XA scale rather well - see [this blog article](https://www.atomikos.com/Blog/JTAXAReliabilityWithoutThePerformanceOverhead).

Since Kafka is already a "distributed commit log" I suspect that adding XA support is not that hard - if there is the will to do it. I suppose all that is needed is a sync() flush on storage and a prepared state for transactions in Kafka. We've helped other vendors include XA support and would be happy to help Kafka also.

**Comment by kbastani** [comment]
I’m going to close this thread since it’s become off topic. This is an eventually consistent system. An eventually consistent dual-write is still transactionally atomic with BASE. There is no safety guarantee in any eventually consistent system. I thank everyone for the comments and advice. Please feel free to make a pull request if you know a better way. Or at least explain clearly the solution in code rather than methodology.

**Comment by thekalinga** [comment]
@kbastani

>I’m going to close this thread since it’s become off topic.
I am not sure if this is completely true as all the threads are attempting to add more context/present ideas on how to solve the issue. Can you plz keep this thread open for the time being as it has opened the can of worms & only thru discussion we will be able to find a better alternative to this fundamental problem

> I thank everyone for the comments and advice. Please feel free to make a pull request if you know a better way. Or at least explain clearly the solution in code rather than methodology.

We need a way to collaborate so that we can come up with a solution that is a reasonable compromise between complexity & performance. Even if we are not able to comeup with the solution rightaway, someone other than us can look at this thread and contribute a solution if the thread is open

> This is an eventually consistent system. An eventually consistent dual-write is still transactionally atomic within the scope of BASE.

The current implementation is broken, so can never be eventually consistent in the scenario we all agreed as broken. Eventual consistency implies the system will converge after a while. But our current implementation can never reach a consistent state even after a really really long time, as the data can go missing from db, but will still be present in kafka (plz refer to our earlier discussion)

Thanks in anycase for your effort in creating this base platform

**Comment by kbastani** [comment]
@thekalinga The issue tracker is for issues, so I'm happy to open up another issue that describes the exact conditions of the defect so that it can be both tested at scale and then remedied. 

I am all about continuing this conversation, but it would be a repeat of the same conversation I've had over and over for years now. You're correct that the current implementation is not atomic. Eventually consistent transactions use BASE, which means that there is no way to implement atomicity in a distributed transaction without sacrificing performance.

As for the likelihood of an inconsistency arising from an application crash, there are many assumptions being made here that are not considering the operating system, the JVM, and the container scheduler (k8). So I'm happy to explain why I closed this thread.

In the case that the application experiences a SIGKILL signal while transactions are in-flight, it could only occur if and only if the application is unresponsive to a SIGTERM signal sent by the Linux kernel before the scheduler sends a system shutdown command to the container. The likelihood of that happening while operating the application on something like Kubernetes is very very unlikely. When a container crashes on Kubernetes due to something like a failed health check, a SIGKILL is not sent to the JVM application immediately. The difference here is that a SIGKILL is like a CTRL+ALT+DELETE way to terminate a process in a POSIX-compliant Unix operating system. While SIGTERM is a much friendlier way to notify a process that it should finish up its work and shutdown gracefully before it gets killed.

What this means is, when a Spring Boot application receives a SIGTERM signal from the Linux kernel due to a critical fault...if the application is responsive...it will shutdown gracefully after all in-flight transactions are finalized. If the application does not shutdown in a certain period of time, then the Kernel will then send a SIGKILL, which would then immediately terminate the JVM process. Should the application become unresponsive, that is only likely to happen if the JVM heap is exceeded. Because we're cloud native and use horizontal auto-scaling, the likelihood of the application becoming unresponsive to a SIGTERM signal is just about 99.99999% improbable. 

Now let's talk through other scenarios that could possibly terminate the application in the middle of a dual-write... The only other offender would be the nefarious segfault. Well I'm happy to say that the JVM cannot experience a segfault unless someone was doing something very silly... like creating a pointer to an unsafe memory space using a native system call. None of this is likely to happen in an opinionated framework like Spring Boot.

So yes, we are debating an age old argument that is disregarding many of the most salient points of what happens when the application crashes. I'm happy to allow open discussion here, but I would urge you to defer to Chris Richardson's book, which has the best possible advice on this very topic.

**Comment by GuyPardon** [comment]
Thanks, but just my 2 cents: 

1. Out-of-memory errors are a typical example of exceptions you cannot foresee nor catch. They lead to transaction recovery errors not solved by any of the replies in this thread. Plus they happen at the worst time: when the number of transactions in flight is at a maximum, thereby maximising the damage being done. You probably want XA recovery in that case.

2. Being a long-time advocate and co-inventor of BASE / compensation / saga solutions, I have to say that we discovered that it is mostly a myth if you account for disaster recovery. You can't recover in a consistent way with those. Chris Richardson can't help either AFAIK. If you're interested, just google for "The BAC theorem" to learn more. We've even come to conclude that it's not "eventually consistent" but rather "eventually IN-consistent" for that matter. 

Disclaimer: I am co-author of the research involved :-) Being the geeks we are, we're working on a solution, and only XA transaction recovery looks like it could offer some hope so far... 

So it's of course up to you to close this discussion but I agree with the others that the issue is not solved by any of the solutions proposed.

Sincerely yours,
Guy

**Comment by thekalinga** [comment]
@kbastani Thanks for your detailed response

I am okay to open a separate thread to continue this conversation

I think you have not accounted for few more scenarios

1. Machine failure (with extremely low probablity, but if you have hundreds of machines & millions of transactions, the likely hood goes up significantly). Machine crashing is not an unusual occurrence
2. What if you have a dead lock/live lock in the RDBMS & the commit has to wait or the transaction needs to be rolledback after kafka tx is complete

I am sure there are other scenarios that I am missing as I only have limited knowledge

In anycase, it would help all of us if we have a thread where this discussion can continue

Thanks again

**Comment by kbastani** [comment]
Well now I’m intrigued. The only thing here that might need to be
reassessed in your research is that reactive streams do not block during a
transaction. So the number of in-flight transactions being maxed out may
affect your analysis.

If it’s database cursors that cause the transaction limit to be exceeded
then the TX fails and no harm done.

Can we create a working group on this subject and update your research?

**Comment by thekalinga** [comment]
@GuyPardon Sorry 2PC quite a lot of downsides. The following article should gives decent justifications on why its a performance killer

https://dbmsmusings.blogspot.com/2019/01/its-time-to-move-on-from-two-phase.html

**Comment by kbastani** [comment]
I really appreciate your willingness to bring up more potential blind
spots. I suggest we create a working group with a weekly 30-minute
conference call to definitively prove the likelihood of an inconsistency
occurring on K8.

Wdyt?

**Comment by thekalinga** [comment]
@kbastani 
> If it’s database cursors that cause the transaction limit to be exceeded then the TX fails and no harm done.

But our kafka is already commited by then

Here is our flow

1. DB TX starts
2. Make DB changes
3. Kafka TX starts
4. Send message to Kafka
5. Kafka TX commit
6. DB TX deadlock/live lock
7. DB rollback

As you can see Only DB TX is rolled back, but not Kafka

**Comment by kbastani** [comment]
@thekalinga That’s actually not entirely true. Please refer to the test I referenced earlier in this thread. It proves that if the database transaction fails that the Kafka event is never sent.

**Comment by thekalinga** [comment]
@kbastani That's a nice idea. I'm more than happy to join the conversation

Would be great if others are also interested in proposing some solutions to solve this problem thru people who are interested in coming up with multiple alternative implementations in solving this problem, 

> @thekalinga That’s actually not entirely true. Please refer to the test I referenced earlier in this thread. It proves that if the database transaction fails that the Kafka event is never sent

Sure, I'll check it

One reason for this might be because kafka batches your requests to achieve better throughput with default settings. If you are unlucky, the messages will be dropped incase of a crash. Again, I am not an expert, this is only a guess based on transaction guarantees of any store

**Comment by kbastani** [comment]
@thekalinga Great. I’ll make an open call for researchers interested in
establishing a working group over a period of no more than 5 weeks.

I strongly urge you to explore how Reactor completely changes everything we
know about what is possible when the database, application, and API gateway
are all reactive and non-blocking. I’ve designed the system to explore
which constraints around transactions are nullified due to the unique
communication protocol that undoubtedly changes the transactional scope of
any operation. For instance, I can rollback a transaction from the API
gateway by talking directly to the database. It sounds bizarre but that’s
why this project exists.

**Comment by thekalinga** [comment]
@kbastani 

> I strongly urge you to explore how Reactor completely changes everything we know about what is possible when the database, application, and API gateway are all reactive and non-blocking....

:) I am a contributor to reactor. I have non-noob amount of code level understanding of reactive streams & reactor. I along with @simonbasle have redesigned almost all the marble diagrams, for which I had to look at the implementation of some reactor operators. Which is why I said before that reactive does not change the underlying transnational semantics from the external data store point of view

In anycase, I'll go over the test and confirm

Thanks for your initiative

Better to create an agenda before we meet & do some background reading. It would be great If we can find people who have already gone thru the pains of CQRS before & are willing to share/propose a new/better solutions

1. Compensating transactions (SAGA)
2. Process manager/complex event processing
3. Change Data Capture (CDC)
4. 2PC
5. May be other solutions

Some sources that might be of help

* https://medium.com/netflix-techblog/scaling-event-sourcing-for-netflix-downloads-episode-1-6bc1595c5595
* https://medium.com/netflix-techblog/scaling-event-sourcing-for-netflix-downloads-episode-2-ce1b54d46eec

**Comment by kbastani** [comment]
@thekalinga Let’s please try to leave our egos, pedigrees, and accomplishments at the door here. My statement stands as stated whether or not you’re an expert or a novice. I’ll follow up soon with a new issue and get contacts exchanged. Thank you so much for participating in this project and adding your valuable input.

**Comment by thekalinga** [comment]
>Let’s please try to leave our egos, pedigrees, and accomplishments at the door here.
Sure..

Will share my email address with you over twitter

**Comment by thekalinga** [comment]
I just came across the baby of @cer. I'll go over it before we have round of discussion

May be we need to adopt it to Kafka as a first step

https://github.com/crctraining/customers-accounts-and-money-transfers

**Comment by thekalinga** [comment]
@GuyPardon Wouldn't we get consistent backups if we follow @nurkiewicz solution of kafka as single source of truth?

@nurkiewicz 

> Any solution at the API level (reactive, callback-base, etc.) can overcome the issues with underlying network protocol.

I believe you meant "Any solution at the API level (reactive, callback-base, etc.) **cannot** overcome the issues with underlying network protocol."

Also wanted to check with you how you handle the out of sync db views when you need to apply business constraints. Classic example: Unique email id, money transfer from one account to next

Should we always live with compensating transactions? But I am not fully satisfied with the eventual aspect for cases like unique email id. Do you have practical experience in this area?

But one pattern I liked is the netflix one where they use casandra as both event store & also as snapshot store with in memory aggregate models

But the only thing is for each request we have replay events from last snapshot which feels like a lot of work

I know there is decent amount of pain, but is the gain really worth it (or) are there any ways which can ease this pain slightly

**Comment by cer** [comment]
The key pattern that guarantees atomicity is
https://microservices.io/patterns/data/application-events.html aka
transactional outbox

**Comment by thekalinga** [comment]
@cer Thanks for your response

Yes this can for sure ease the pain

I assume we would want to keep this relay code separately (something like CDC) and not keep this as part of the application as multiple instance of our application will result in duplicate entries.

Is this true?

**Comment by cer** [comment]
Yes. http://eventuate.io has a CDC service (standalone application) thats responsible for publishing messages/events inserted into OUTBOX table to a message broker.

**Comment by GuyPardon** [comment]
Hi,

@thekalinga Yes, if you use one single source of truth then it would work (whether Kafka or something else for that matter).

We are trying to find a way to make work without a single source of truth and that seems to require XA, but I may be biased.

Cheers,
Guy

**Comment by gunnarmorling** [comment]
I'd recommend to follow the of @cer and go for an CDC-based outbox approach.

Even if you could somehow make Kafka XA-capable (or employ something like the "last resource gambit"), I don't think it's desirable, as it ties your service and its readiness to the availability of a remote messaging broker. Outbox avoids that by processing the events from the outbox table asynchronously, so temporarily not being able to emit your message towards Kafka won't impact your own service's availability.

Funnily enough, we just did a [post](https://debezium.io/blog/2019/02/19/reliable-microservices-data-exchange-with-the-outbox-pattern/) about this very use case of ensuring eventual consistency between a DB and Kafka on the Debezium blog. Perhaps it can serve as input to drive the solution for this issue here?

**Comment by MrKnyaz** [comment]
Every Saga starts with an event. Don't try to write to multiple systems in the first step of your Saga)
Its ok to do this on second and other steps, as in the worst case scenario you'll have to reprocess message again and think of idempotency, but never lose an event. (of course if you ack/commit message after processing(saving something to DB and sending msg to next queue/topic) and not before)
